### ISSUE TITLE
Synology DSM issues reporting service status

### ISSUE THREAD

**Issue by adhodgson1** [nextdns/nextdns#413]
NextDNS 1.10.1 installed on Synology DSM 6.2.3 with no issues and everything working as expected. However I am seeing issues with reporting service status on a running system as follows:

admin@frodo:/$ sudo nextdns status
Error: /usr/local/etc/rc.d/nextdns.sh status: exit status 1:
admin@frodo:/$ sudo /usr/local/etc/rc.d/nextdns.sh status
Stopped

I found this issue as I wanted to make a change to the configuration file for DNS discovery and ended up having to reboot the unit in order to get the service to start correctly, which it did on a reboot.

**Context**
 - Version: 1.10.1
 - Platform: Synology (DSM)

**Comment by jbehrmusic** [comment]
I can confirm this still is an issue. Here's what I'm running.
- DSM 6.2.3-25426 Update 3
- DS218j [hardware]

When I SSH into the NAS, and try to run `nextdns status` , I get this:

~# nextdns status
Error: /usr/local/etc/rc.d/nextdns.sh status: exit status 1:

**Comment by adhodgson1** [comment]
Something else I noticed was that when using nextdns log I get logs from an old nextdns command I ran a few months ago even though a daemon is currently running.

**Comment by jbehrmusic** [comment]
> Something else I noticed was that when using nextdns log I get logs from an old nextdns command I ran a few months ago even though a daemon is currently running.

Yes I noticed this as well. When I restarted my NAS, and ran `nextdns log` , it still showed old log errors from the day before. I was confused at first, but luckily it shows the time/date.

**Comment by threeonparfive** [comment]
> I can confirm this still is an issue. Here's what I'm running.
> 
> * DSM 6.2.3-25426 Update 3
> * DS218j [hardware]
> 
> When I SSH into the NAS, and try to run `nextdns status` , I get this:
> 
> ~# nextdns status
> Error: /usr/local/etc/rc.d/nextdns.sh status: exit status 1:

Same issue on a DS220j running same versions.

**Comment by rs** [comment]
@adhodgson1 Can you please send a PR?

**Comment by adhodgson1** [comment]
Check the file it may have been overwritten with the template in the binary. I eventually replaced the binary on my unit with the one I built myself with the updated template as I came up against this issue.

**Comment by xyzulu** [comment]
I'm still seeing issues here.. on the latest release.
`SynologyRouter> nextdns status
Error: /usr/local/etc/rc.d/nextdns.sh status: exit status 1: `

Yet nextdns is running just fine.. I can see it in ps as well as query it on port 53
I'm using this on SRM MR2200AC to be exact.

**Comment by adhodgson1** [comment]
Hmm I don’t have a Synology router at this end to play with. Have you tried reverting to v1.11.0 which is the version before I made my change to see if this reports status successfully?

**Comment by xyzulu** [comment]
No, but I could revert your change manually in the template file fairly easily. Good idea.. I'll report back.

**Comment by xyzulu** [comment]
Yip, reverting to:
`is_running() {
	test -f "$pid_file" && ps | grep -q "^ *$(get_pid) "
`
.. fixed it for me. Perhaps SRM and DSM need separate treatment here..

**Comment by xyzulu** [comment]
@rs can you help? ;)

**Comment by adhodgson1** [comment]
Is the SRM running Busybox as default shell or Bash? When I made my original change in March I did ask whether we would hit some cases as I thought that some Synology devices were running Busybox.

**Comment by xyzulu** [comment]
Busybox (ash)

**Comment by adhodgson1** [comment]
I've added a quick check in this [pull request](https://github.com/nextdns/nextdns/pull/486) to see if we're running under Busybox. I am running this version on my DSM based NAS, and tested the relevant code under a couple of Busybox containers. @xyzulu can you check it on your SRM device and report back?

**Comment by xyzulu** [comment]
Looks good..  Olivier has merged it already. I'll test the whole release (uninstall / install) again when it is released.
Thanks!

**Comment by eizedev** [comment]
I have also the same problem on my synology router. On my two synology NAS systems the current release is working so i think the fix applied by @adhodgson1 will work (tried the discovery if nextdns is running manually on my router). Waiting for the next release @rs :-)

Thanks

**Comment by xyzulu** [comment]
The latest SRM update (from a few days ago) has broken nextdns on SRM (Synology router) badly now.. but I don't see issues related to Nextdns SRM documentation I have opened here on Github being taken care of. Not sure what is going on...

**Comment by eizedev** [comment]
Infos for debugging @rs and @adhodgson1 
On my router i can see that the service is running, but the socket was not created (`/var/run/nextdns.sock`).
This was related to two nextdns processes that were started:

```bash
cat /var/run/nextdns.pid
25296
```

```bash
pidof nextdns
30594 23059
```

`test -f "/var/run/nextdns.pid" && ps | grep -q "^ *25296 "`

```bash
echo $?
1
```

```bash
ps -w | grep nextdns
23059 root      782m S    /volume1/@spool/_airprint_drivers/bin/nextdns run
30594 root      782m S    /volume1/@spool/_airprint_drivers/bin/nextdns run
```

After trying to stop these two processes (which did not work since the pid in the pidfile is the wrong one) i killed the two processes, started nextdns again:

![image](https://user-images.githubusercontent.com/6794362/118402945-77f1f180-b66c-11eb-961e-8216e7b5e718.png)
![image](https://user-images.githubusercontent.com/6794362/118403235-d370af00-b66d-11eb-8d34-48fc91040e95.png)


```bash
netstat -tulpn | grep :53 | grep -v avahi | grep -v miniupnpd
tcp        0      0 ::%6:53                 :::*                    LISTEN      8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0   1216 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 0.0.0.0:5353            0.0.0.0:*                           8669/nextdns
udp        0      0 ::%27:5353              :::*                                8669/nextdns
udp        0      0 ::%28:5353              :::*                                8669/nextdns
udp        0      0 ::%29:5353              :::*                                8669/nextdns
udp        0      0 ::%30:5353              :::*                                8669/nextdns
udp        0      0 ::%31:5353              :::*                                8669/nextdns
udp        0      0 ::%32:5353              :::*                                8669/nextdns
udp        0      0 ::%33:5353              :::*                                8669/nextdns
udp        0      0 ::%34:5353              :::*                                8669/nextdns
udp        0      0 ::%35:5353              :::*                                8669/nextdns
udp        0      0 ::%36:5353              :::*                                8669/nextdns
udp        0      0 ::%37:5353              :::*                                8669/nextdns
udp        0      0 ::%38:5353              :::*                                8669/nextdns
udp        0      0 ::%39:5353              :::*                                8669/nextdns
udp        0      0 ::%40:5353              :::*                                8669/nextdns
udp        0      0 ::%41:5353              :::*                                8669/nextdns
udp        0      0 ::%42:5353              :::*                                8669/nextdns
udp        0      0 ::%43:5353              :::*                                8669/nextdns
udp        0      0 ::%44:5353              :::*                                8669/nextdns
udp        0      0 ::%45:5353              :::*                                8669/nextdns
udp        0      0 ::%46:5353              :::*                                8669/nextdns
udp        0      0 ::%47:5353              :::*                                8669/nextdns
udp        0      0 ::%51:53                :::*                                8669/nextdns
```

DHCP Server log after running `nextdns start`

```bash
May 16 17:27:17 RouterBox dhcp-server: Stopped
May 16 17:27:17 RouterBox dnsmasq[8220]: failed to create listening socket for port 53: Address already in use
May 16 17:27:17 RouterBox dnsmasq[8220]: FAILED to start up
May 16 17:27:17 RouterBox [11015.089119] init: dhcpserver main process (8220) terminated with status 2
May 16 17:27:17 RouterBox [11015.089291] init: dhcpserver faild on spawn stage, stopped
May 16 17:27:17 RouterBox dhcp-server: Stopped
May 16 17:27:22 RouterBox dhcp-server: Started
May 16 17:27:29 RouterBox dhcp-server: Stopped
May 16 17:27:29 RouterBox dhcp-server: Started
```

![image](https://user-images.githubusercontent.com/6794362/118403195-986e7b80-b66d-11eb-81e0-3840d9ceabfc.png)

**The isrunning function should work, but i don't know why, it is not working. I can execute the commands manually as you can see.

Additional info:

`cat /etc/resolv.conf`

```bash
# This file is managed by nextdns.
#
# Run "nextdns deactivate" to restore previous configuration.

nameserver 127.0.0.1
```

`/etc/dhcpd/dhcpd.conf`

```bash
cat /etc/dhcpd/dhcpd.conf | grep -iA3 next
# Configuration generated by NextDNS
# DNS is handled by NextDNS
port=0
dhcp-leasefile=/etc/dhcpd/dhcpd.conf.leases
dhcp-script=/usr/syno/etc/dhcpd/dhcpd-script.sh
```

`nextdns config`

```bash
report-client-info true
listen localhost:53
control /var/run/nextdns.sock
cache-max-age 0s
max-ttl 5s
hardened-privacy false
bogus-priv true
config ******
discovery-dns
use-hosts true
timeout 5s
log-queries false
cache-size 10MB
detect-captive-portals false
setup-router true
auto-activate true
```

`netstat -tulpn | grep 53`

```bash
tcp        0      0 127.0.0.1:5342          0.0.0.0:*               LISTEN      30594/nextdns
tcp        0      0 ::%6:53                 :::*                    LISTEN      23059/nextdns
```

```bash
readlink /bin/ls 2>&1 | grep busybox
busybox
```

**Comment by adhodgson1** [comment]
I'm confused about whether this issue is due to the SRM update discussed by @xyzulu or the fact you are running a version of NextDNS which hasn't taken in the merged changes. If you are running the latest release 1.32.1 then that hasn't got the latest change which I made last week to cope with the SRM Busybox case which is used to check running status. If you want a version without my original change then try running with version 1.10.1. @xyzulu confirmed this was working for him before he upgraded to the new SRM release last week. I think anything else from herein needs to go in a new issue as this one has received a couple of merges already and was created for the DSM case. I can only offer limited info on the SRM because I don't own the equipment. I just put in this feature for SRM as it effectively restored the original behaviour for SRM if its detected we're running under Busybox which I think will be used on some lower spec DSM equipment anyhow. Closing for now.
